File: javascript/food-chain/food-chain.js
var FoodChain = function() {

  var CHAIN = {
    '': '',
    'fly': "I don't know why she swallowed the fly.",
    'spider': "It wriggled and jiggled and tickled inside her.",
    'bird': "How absurd to swallow a bird!",
    'cat': "Imagine that, to swallow a cat!",
    'dog': "What a hog, to swallow a dog!",
    'goat': "Just opened her throat and swallowed a goat!",
    'cow': "I don't know how she swallowed a cow!",
    'horse': "She's dead, of course!"
  };

  var chain = function(i) {
    var animal = Object.keys(CHAIN)[i];
    if (i < 2) return '';
    return "She swallowed the " + animal +
      " to catch the " + fly(i-1) + ".\n" +
      chain(i-1);
  };

  var fly = function(i) {
    var animal = Object.keys(CHAIN)[i];
    if (i == 2) animal += ' ' + CHAIN[animal].replace(/It/,'that').replace(/\./,'');
    return animal
  };

  this.verse = function(i) {
    var animal = Object.keys(CHAIN)[i];
    var out = [];
    out.push("I know an old lady who swallowed a " + animal + ".\n");
    if (i > 1) out.push(CHAIN[animal] + "\n");
    if (i < 8) {
      out.push(chain(i));
      out.push(CHAIN['fly'] + " Perhaps she'll die.\n");
    }
    return out.join('')
  };

  this.verses = function(start, stop) {
    var out = [];
    for (var i = start; i <= stop; i++) {
      out .push(this.verse(i));
    }
    return out.join("\n") + "\n";
  };

};

module.exports = FoodChain;

File: ruby/food-chain/food_chain.rb
module BookKeeping
  VERSION = 2
end

module FoodChain

  def self.song
    CHAIN.keys.map.with_index do |_animal, i|
      self.verse(i)
    end.join("\n\n")
  end

  CHAIN = {
    'fly' => "I don't know why she swallowed the fly.",
    'spider' => "It wriggled and jiggled and tickled inside her.",
    'bird' => "How absurd to swallow a bird!",
    'cat' => "Imagine that, to swallow a cat!",
    'dog' => "What a hog, to swallow a dog!",
    'goat' => "Just opened her throat and swallowed a goat!",
    'cow' => "I don't know how she swallowed a cow!",
    'horse' => "She's dead, of course!"
  }

  def self.verse(i)
    animal = CHAIN.keys[i]
    [].tap do |out|
      out << "I know an old lady who swallowed a #{animal}.\n"
      out << "#{CHAIN[animal]}\n" unless animal == 'fly'
      unless animal == 'horse'
        out << self.chain(i)
        out << "#{CHAIN['fly']} Perhaps she'll die."
      end
    end.join
  end

  def self.chain(i)
    return '' if i < 1
    spider = CHAIN.keys[i]
    "She swallowed the #{spider} to catch the #{self.fly(i-1)}.\n" + self.chain(i-1)
  end

  def self.fly(i)
    animal = CHAIN.keys[i]
    [].tap do |out|
      out << animal
      if animal == 'spider'
        out << CHAIN[animal].sub('It','that').sub('.','')
      end
    end.join(' ')
  end

end
